lib/services/unified_resource.go

```diff
@@ -193,127 +196,327 @@ func (c *UnifiedResourceCache) deleteLocked(res types.Resource) error {
+		for item, err := range c.iterateItems(ctx, start, sortBy, kinds...) {
+			if err != nil {
+				yield(nil, err)
+				return
 			}
 
-			// do we have all we need? set nextKey and stop iterating
-			// we do this after the matchFn to make sure they have access to the "next" node
-			if req.Limit > 0 && len(res) >= int(req.Limit) {
-				nextKey = item.Key.String()
-				return false
+			if !yield(item.resource.CloneResource(), nil) {
+				return
 			}
-			res = append(res, resourceFromMap)
-			return true
-		})
-		return iteratorErr
-	})
-	if err != nil {
-		return nil, "", trace.Wrap(err)
+		}
 	}
+}
+
+// IterationOrder dictates the order in which iteration should
+// return items.
+type IterationOrder int
+
+const (
+	// IterateAscend indicates that items should be returned in ascending order.
+	IterateAscend IterationOrder = iota
+	// IterateDescend indicates that items should be returned in descending order.
+	IterateDescend
+)
+
+// IterateNodes iterates over all cached nodes starting from the provided key.
+func (c *UnifiedResourceCache) IterateNodes(ctx context.Context, start string, order IterationOrder) iter.Seq2[types.Server, error] {
+	return iterateUnifiedResourceCache(ctx, c, start, order, types.KindNode, types.Server.DeepCopy)
+}
+
+// IterateAppServers iterates over all cached app servers starting from the provided key.
+func (c *UnifiedResourceCache) IterateAppServers(ctx context.Context, start string, order IterationOrder) iter.Seq2[types.AppServer, error] {
+	return iterateUnifiedResourceCache(ctx, c, start, order, types.KindAppServer, types.AppServer.Copy)
+}
+
+// IterateDatabaseServers iterates over all cached database servers starting from the provided key.
+func (c *UnifiedResourceCache) IterateDatabaseServers(ctx context.Context, start string, order IterationOrder) iter.Seq2[types.DatabaseServer, error] {
+	return iterateUnifiedResourceCache(ctx, c, start, order, types.KindDatabaseServer, types.DatabaseServer.Copy)
+}
+
+// IterateKubernetesServers iterates over all cached Kubernetes servers starting from the provided key.
+func (c *UnifiedResourceCache) IterateKubernetesServers(ctx context.Context, start string, order IterationOrder) iter.Seq2[types.KubeServer, error] {
+	return iterateUnifiedResourceCache(ctx, c, start, order, types.KindKubeServer, types.KubeServer.Copy)
+}
 
-	if len(res) == backend.DefaultRangeLimit {
-		c.logger.WarnContext(ctx, "Range query hit backend limit. (this is a bug!)",
-			"start_key", startKey,
-			"range_limit", backend.DefaultRangeLimit,
-		)
+// IterateWindowsDesktops iterates over all cached windows desktops starting from the provided key.
+func (c *UnifiedResourceCache) IterateWindowsDesktops(ctx context.Context, start string, order IterationOrder) iter.Seq2[types.WindowsDesktop, error] {
+	return iterateUnifiedResourceCache(ctx, c, start, order, types.KindWindowsDesktop, func(desktop types.WindowsDesktop) types.WindowsDesktop { return desktop.Copy() })
```

Comment:
why cant this copy func just be `types.WindowsDesktop.Copy` like the other types?

Reply:
ah, guess i could have just tried myself :)

LGTM!